### ISSUE TITLE
Make modes and means optional

### ISSUE THREAD

**Issue by raphaelshirley** [LSSTDESC/rail_bpz#43]
** Feature request**

The BPZ estimator is returning the posterior modes and means. I think it may be recommended that point estimators are optional because of computational expense.

See e.g. line:

https://github.com/LSSTDESC/rail_bpz/blob/main/src/rail/estimation/algos/bpz_lite.py#L578

Can we make the point estimators optional?

**Before submitting**
Please check the following:

- [x] I have described the purpose of the suggested change, specifying what I need the enhancement to accomplish, i.e. what problem it solves.
- [x] I have included any relevant links, screenshots, environment information, and data relevant to implementing the requested feature, as well as pseudocode for how I want to access the new functionality.
- [x] If I have ideas for how the new feature could be implemented, I have provided explanations and/or pseudocode and/or task lists for the steps.

**Comment by eacharles** [comment]
I think there's a bit more to it than that.

So, the line in question is:
`qp_dstn.set_ancil(dict(zmode=zmode, zmean=zmean, tb=tb, todds=todds))`

Which is just setting the already calculate values in the ancillary data.
zmean is computed a few lines above, in: 
`zmean[i] = (zgrid * pdfs[i]).sum() / pdfs[i].sum()`

And `zmode`, `tb`, and `todds` are computed as part of self._estimate_pdf()

I had a closer look and it seems like tb and todds give information about the template choice and aren't really point estimates.
And zmode is used to compute `tb` and `todds`

So, 

1. It would be easy to turn off zmean, but I doubt that is taking up much time
2. We could not include zmode, tb and todds, but that just saves space, not time, unless we turn off the computation.
3. Zmode is probably the slowest of the three, the other two just look like sums / pulling out values.

Do we have evidences that this is really slowing things down by a substantial amount (say more that 5% relative to the computation of the pdf itself)?

**Comment by raphaelshirley** [comment]
This was based on an original discussion with @aimalz Perhaps the point estimates are needed specifically for BPZ so should remain?

**Comment by eacharles** [comment]
Well, the only thing that isn't BPZ specific is the mean.   I doubt that is taking up much time at all.  

We could provide an option to turn off the combination of zmode, tb, and todds, but those seem kinda like something that is generally useful for bpz, so I would not make that the default.  

So, barring evidence that computing zmode, tb and todds is slowing bpz down appreciably, my vote would be to do nothing at this time.

**Comment by sschmidt23** [comment]
As Eric said, you get the `zmode` "for free" with BPZ because the algorithm computes the PDF on the input grid for every galaxy in the course of computing the likelihoods and posteriors.  So, it's literally just a single np.argmax command here: https://github.com/LSSTDESC/rail_bpz/blob/2d51caa261984467513608d158d9745b36bababc/src/rail/estimation/algos/bpz_lite.py#L512

tb was a feature request by someone, I can't remember who, and it's practically free, it's a simple check of elements of an already computed array, it's an argmax over the other direction of the array: https://github.com/LSSTDESC/rail_bpz/blob/2d51caa261984467513608d158d9745b36bababc/src/rail/estimation/algos/bpz_lite.py#L530

So, those two would be trivial amounts of compute time to not compute, and I think that's why we just write them out.  

If we wanted to make those optional we could swap in code like that in e.g. rail_flexzboost with `calculated_point_estimates` here: https://github.com/LSSTDESC/rail_flexzboost/blob/b4ac032add2971018bb45b661642f94fdc656171/src/rail/estimation/algos/flexzboost.py#L243  (though without the if statements, since we know it is an interp distribution for BPZ), which should be fairly trivial.

Maybe we could create `save_tb` and `save_todds` flags (with defaults of `True` so you have to actively turn them off if you don't want to save) to not write those out?

**Comment by drewoldag** [comment]
I don't want to wade too deeply into a debate, but I agree that since zmode is "for free" it totally makes sense to keep it. But I get the sense that the same line of reasoning that we've applied to "don't do work in the `__init__` method" probably applies here as well. i.e. if the user isn't explicitly asking for the mean, then perhaps it's not worth calculating 🤷 I certainly don't know the typical use cases though, maybe the general user expects the mean to be there by default. 

Also, we did the work to abstract the point estimation logic out using the `PointEstimateMixin` [here](https://github.com/LSSTDESC/rail_base/blob/main/src/rail/core/point_estimation.py). in an attempt to consolidate the point estimation code to ease maintenance. But again, not knowing the details of bpz, perhaps it's very efficient to get the point estimates due to some earlier calculations.

**Comment by eacharles** [comment]
Yes, we could use the PointEstimateMixin, but I since we are already computing the mode in `self._estimate_pdf()` requesting the mode would re-do that computation unless we put in extra stuff to catch that particular case.

**Comment by sschmidt23** [comment]
I think the re-computation may be why I didn't update BPZ to use the mixin class (or I just forgot that we never addressed point-estimate computation it in this repo).  Whatever the case, we can definitely refactor the code here to allow for more customization by the user, as right now the ancil data returned is all hardcoded.

I seem to have a vague recollection of discussions of custom methods for point estimates when we knew they were efficient (e.g. mean calculation for mixmod being fairly simple, and cases where the  qp.interp grid is the same as the native storage grid for mode), though I think that discussion was for code in `qp`, and I think some of the custom functions may have been removed during a refactoring.  We can discuss what to do in terms of efficiency so that we're not duplicating work to re-compute the mode in the code.

**Comment by drewoldag** [comment]
Yes, I also recall the conversations about taking advantage of highly efficient code if it's available. And I'm pretty sure we had that conversation about zmode here as an example of where it makes sense _not_ to use the mixin. So along the same lines, if the other point estimates are just as fast, then yeah, it total makes sense to keep things as they are (or add some commants about why they are the way they are)

**Comment by eacharles** [comment]
This has been upstreamed to the base class.
